sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,325 @@
+    {
+        use std::io::Read;
+        use std::io::Write;
+        let mut port = serialport::new(self.console_tty.to_str().ok_or(Error::UnicodePathError)?, 115_200)
+            .timeout(std::time::Duration::from_millis(100))
+            .open().expect("Failed to open port");
+
+        // Ideally, we would invoke Linux flock() on the serial
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf: [u8; 128] = [0; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    print!("Discarded {} characters: {:?}\n", rc, &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            break;
+                        }
+                        _ => {
+                            print!("Error reading: {:?}\n", error);
+                            return Err(error.into())
+                        }
+                    }
+                }
+            }
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
+        let mut seen_echo = false;
+        let mut len: usize = 0;
+        loop {
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
+                        for i in 0..len {
+                            if buf[i] == 10 {
+                                // Newline
+                                let mut end = i;
+                                if end > 0 && buf[end - 1] == 13 {
+                                    end -= 1;
+                                }
+                                let line = std::str::from_utf8(&buf[0..end])?;
+                                if seen_echo {
+                                    callback(line);
+                                } else {
+                                    let minmatch = std::cmp::min(6, cmd.len());
+                                    if line.len() >= minmatch && line[line.len() - minmatch..]
+                                        == cmd[cmd.len() - minmatch..] {
+                                        seen_echo = true;
+                                    }
+                                }
+                                buf.rotate_left(i + 1);
+                                len -= i + 1;
+                                continue 'outer;
+                            }
+                        }
+                        break
+                    }
+                }
+                Err(error) => {
```

Comment:
You might be able to be more expressive with the error conditions:

```
match port.read(...) {
  Ok(rc) => { ... }
  Err(e) if e.kind == ErrorKind::TimedOut => { ... }
  Err(e) => return Err(e)
}
```

Reply:
Thanks for the suggestion, doing so does save a couple of lines, and two levels of indentation.